api/v1/egressgateway_types.go

```diff
@@ -0,0 +1,277 @@
+}
+
+// EgressGatewayDeploymentPodTemplateSpec is the EGW Deployment's PodTemplateSpec
+type EgressGatewayDeploymentPodTemplateSpec struct {
+	// Metadata is a subset of a Kubernetes object's metadata that is added to
+	// the pod's metadata.
+	// +optional
+	Metadata *EgressGatewayMetadata `json:"metadata,omitempty"`
+
+	// Spec is the EGW Deployment's PodSpec.
+	// +optional
+	Spec *EgressGatewayDeploymentPodSpec `json:"spec,omitempty"`
+}
+
+// EgressGatewayMetadata contains the standard Kubernetes labels and annotations fields.
+type EgressGatewayMetadata struct {
+	// Labels is a map of string keys and values that may match replica set and
+	// service selectors. Each of these key/value pairs are added to the
+	// object's labels provided the key does not already exist in the object's labels.
+	// If not specified will default to projectcalico.org/egw:<name>, where <name> is
+	// the name of the Egress Gateway resource.
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Annotations is a map of arbitrary non-identifying metadata. Each of these
+	// key/value pairs are added to the object's annotations provided the key does not
+	// already exist in the object's annotations.
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type EgressGatewayIPPool struct {
+	// Name is the name of the IPPool that the Egress Gateways can use.
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// CIDR is the IPPool CIDR that the Egress Gateways can use.
+	// +optional
+	CIDR string `json:"cidr,omitempty"`
+}
+
+// NativeIP defines if Egress Gateway pods should have AWS IPs.
+// When NativeIP is enabled, the IPPools should be backed by AWS subnet.
+type NativeIP string
+
+const (
+	NativeIPEnabled  NativeIP = "Enabled"
+	NativeIPDisabled NativeIP = "Disabled"
+)
+
+type LogLevel string
+
+const (
+	LogLevelTrace LogLevel = "Trace"
+	LogLevelInfo  LogLevel = "Info"
+	LogLevelDebug LogLevel = "Debug"
+	LogLevelWarn  LogLevel = "Warn"
+	LogLevelFatal LogLevel = "Fatal"
+	LogLevelError LogLevel = "Error"
+)
+
+// EgressGatewayFailureDetection defines the fields the needed for determining Egress Gateway
+// readiness.
+type EgressGatewayFailureDetection struct {
+
+	// HealthTimeoutDataStoreSeconds defines how long Egress Gateway can fail to connect
+	// to the datastore before reporting not ready.
+	// This value must be greater than 0.
+	// Default: 90
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=2147483647
+	// +optional
+	HealthTimeoutDataStoreSeconds *int32 `json:"healthTimeoutDataStoreSeconds,omitempty"`
+
+	// ICMPProbe define outgoing ICMP probes that Egress Gateway will use to
+	// verify its upstream connection. Egress Gateway will report not ready if all
+	// fail. Timeout must be greater than interval.
+	// +optional
+	ICMPProbe *ICMPProbe `json:"icmpProbe,omitempty"`
```

Comment:
Don't we need one of ICMPProbe / HTTPProbe to be defined? Is it valid to have both of them nil? I would have thought that's an invalid config, and if so we should call it out in the comments.

Reply:
It is valid to have both of them nil.

Reply:
What does it mean when both of them are nil? What is the reason a user would do that?

Reply:
ICMP probes and HTTP probes are used to verify if the EGW can communicate with an external server. This is used to do a health check if EGW pods can reach outside the cluster. I think users can choose not to configure any of the probes and still have EGWs working fine. @fasaxc

Reply:
Yes, exactly. If users want to disable failure detection altogether, then they must leave `EgressGatewayFailureDetection` as nil the main spec.

If the user provides an `EgressGatewayFailureDetection`, but doesn't configure any failure detection modes (ICMP or HTTP) then that is an invalid configuration that doesn't make sense and we cannot action, and so we need to reject it.

For example, the following yaml is invalid and should be rejected: 

```yaml
spec:
  egressGatewayFailureDetection: 
    healthTimeoutDataStoreSeconds: 30
    icmpProbe: nil
    httpProbe: nil
```

Reply:
This is how disabling failure detection should be done:

```yaml
spec:
  egressGatewayFailureDetection: nil
```